design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+
+### Motivation
+
+The current implementation makes the deployment of the Secrets Provider does not support secret rotation and makes the deployment of the Secrets Provider deeply coupled to the customer's applications which not all customers want.
+
+### Requirement
+
+The solution should support the following capabilities:
+
+1. Deployment, deploy as separate Conjur entity; serve K8s Secrets to many Applications
+2. Rotation, update K8s Secret with Conjur secret values
+3. Support removal of deployment
+
+## Solution
+This solution will be divided into milestones.
+
+- Milestone 1: Deployment, deploy as as separate entity
+- Milestone 2: Handle rotation, updating K8s Secrets with Conjur secret value
+
+### Milestone 1: Deployment
+
+#### Design
+
+The Secrets Provider for K8s will no longer run as an *init* container attached to the application pod, rather run as a separate single Job in a namespace. A Job creates a Pod to perform a task and terminates upon completion. Once the Pod is spun up, it will authenticate to Conjur/DAP in the same way it does today, via authn-k8s. The Pod will then fetch the K8s secret denoted with a label and update them with the Conjur secrets they require.
+
+|      | Deployment                  | Elaboration                                                  | Pros                                                         | Cons                                                         |
+| ---- | --------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Run as Job                  | Run as a K8s Job that deploys a Pod that terminates after task completion | - Minimal resources usage<br />- Minimal code changes<br />- Explicit user experience (user expects Job to run to completion quickly) | - "Job" is not an existing application identity granularities. Adding it can take low priority since customers can run the Job using Namespace / Service Account granularities<br /><br />- Deployment will need to be changed (from Job → Pod) for Milestone 2 (support secrets rotation) |
+| 2    | Run as non-terminating  Pod | Run as a Pod that does not terminate automatically after task completion | - Simple to below<br />                                      | - Wastes resources because we aren't updating K8s secrets and up all the time<br /><br />- For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes. (e.g. secrets provider as init-container, conjur authn k8s secrets, any of our clients or Conjur REST API) |
```

Comment:
1. what is `simple to below`?
2. the con regarding segregation of duty is relevant also for the Job option.

Reply:
1. "Simple to deploy" good catch
2. adding